hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/ClosedWithUnhealthyReplicasHandler.java

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Handler to process containers EC which are closed but have some replicas that
+ * are unhealthy.
+ */
+public class ClosedWithUnhealthyReplicasHandler extends AbstractCheck {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(ClosedWithUnhealthyReplicasHandler.class);
+  private final ReplicationManager replicationManager;
+
+  public ClosedWithUnhealthyReplicasHandler(
+      ReplicationManager replicationManager) {
+    this.replicationManager = replicationManager;
+  }
+
+  /**
+   * Handles a closed EC container with unhealthy replicas. Note that if we
+   * reach here, there is no over, under, or mis replication. This handler
+   * will just send commands to delete the unhealthy replicas.
+   *
+   * <p>
+   * Consider the following set of replicas for a closed EC 3-2 container:
+   * Replica Index 1: Closed
+   * Replica Index 2: Closed
+   * Replica Index 3: Closed replica, Unhealthy replica (2 replicas)
+   * Replica Index 4: Closed
+   * Replica Index 5: Closed
+   *
+   * In this case, the unhealthy replica of index 3 should be deleted. This
+   * is not over replication because that unhealthy replica is not available.
+   * </p>
+   * @param request ContainerCheckRequest object representing the container
+   * @return true if this is a closed EC container with unhealthy replicas,
+   * else false
+   */
+  @Override
+  public boolean handle(ContainerCheckRequest request) {
+    ContainerInfo containerInfo = request.getContainerInfo();
+    if (containerInfo.getReplicationType() != HddsProtos.ReplicationType.EC) {
+      return false;
+    }
+    if (containerInfo.getState() != HddsProtos.LifeCycleState.CLOSED) {
+      return false;
+    }
+
+    Set<ContainerReplica> replicas = request.getContainerReplicas();
+    // create a set of indexes that are closed
+    Set<Integer> closedIndexes = replicas.stream()
+        .filter(replica -> replica.getState() == State.CLOSED)
+        .map(ContainerReplica::getReplicaIndex)
+        .collect(Collectors.toSet());
+
+    boolean foundUnhealthy = false;
+    // send delete commands for unhealthy replicas
+    for (ContainerReplica replica : replicas) {
+      if (replica.getState() == State.UNHEALTHY) {
+        /* Sanity check to ensure this index is not under replicated: verify
+        that a closed replica with this index is also present.
+         */
+        if (!closedIndexes.contains(replica.getReplicaIndex())) {
+          LOG.warn("Not handling container {} because replica index {} is " +
+                  "under replicated. Not deleting UNHEALTHY replica [{}].",
+              containerInfo.containerID(), replica.getReplicaIndex(), replica);
+          return false;
+        }
+
+        foundUnhealthy = true;
+        sendDeleteCommand(containerInfo, replica);
+      }
+    }
+
+    // some unhealthy replicas were found so mark UNHEALTHY (note that
+    // UNHEALTHY is currently not a LifeCycleState for a Container)
+    if (foundUnhealthy) {
```

Comment:
Lets leave this as it is for now, but I am not sure about this. To get here, the container is neither over or under replicated, but it has some unhealthy replicas to be removed. Its not really degraded in anyway from a users perspective, except that it uses more space than it should.

With that in mind, we have no distinction between an under-replicated container due to missing replicas vs under-replicated due to unhealthy replicas. From a users perspective they are much the same thing. The only way to debug this, would be to take a missing container, and list its replicas, and from there you should be able to see some are unhealthy.

We will go ahead with this for now, but we may have to have a think about these states, as unhealthy sounds bad, where it isn't really in this case.